default/python/AI/MoveUtilsAI.py

```diff
@@ -164,37 +165,67 @@ def get_nearest_supplied_system(start_system_id):
         return universe_object.System(supply_system_id)
 
 
-def get_nearest_drydock_system_id(start_system_id):
-    """Get system_id of nearest drydock capable of repair.
-
-    This function does not consider starlane obstruction!
+def get_best_drydock_system_id(start_system_id, fleet_id):
+    """
+    Get system_id of best drydock capable of repair, where best is nearest drydock
+    that has a current and target happiness greater than the HAPPINESS _THRESHOLD
+    with a path that is not blockaded or that the fleet can fight through to with
+    acceptable losses.
 
      :param start_system_id: current location of fleet - used to find closest target
      :type start_system_id: int
      :return: closest system_id capable of repairing
      :rtype: int
      """
-    # TODO consider starlane obstruction
+    if start_system_id == -1:
+        print >> sys.stderr, "get_best_drydock_system_id passed bad system id."
+        return None
+
+    if fleet_id == -1:
+        print >> sys.stderr, "get_best_drydock_system_id passed bad fleet id."
+        return None
+
+    HAPPINESS_THRESHOLD = 5
+
     universe = fo.getUniverse()
+    start_sys = universe.getSystem(start_system_id)
     drydock_system_ids = set()
     for sys_id, pids in ColonisationAI.empire_dry_docks.iteritems():
+        if sys_id == -1:
+            print >> sys.stderr, "get_best_drydock_system_id passed bad dry dock sys_id."
+            continue
         for pid in pids:
             planet = universe.getPlanet(pid)
-            if planet and planet.currentMeterValue(fo.meterType.happiness) >= 5:
+            if (planet
+                  and planet.currentMeterValue(fo.meterType.happiness) >= HAPPINESS_THRESHOLD
```

Comment:
Ideally, we check happiness at the expected arrival time and for the expected duration of the repairs.  

This version means the AI targets planets that are happy now and will be happy when it arrives.

It doesn't get trapped sitting at a planet that will be happy enough in 5 turns, but for now is a death trap.